[9.0] fix handling of *2many field values coming from an onchange

rco-odoo opened the pull request:
When an onchange method adds a record in a many2many field, it is returned to the client as an "update" code `[(1, id, values)]`, where `values` contains the values of the fields shown in the view.  Returning a code `1` (update) instead of a code `4` (link) has the advantage to avoid the client return extra queries to read the content of the added line.

The problem is: writing on a many2many with an update code performs the update *only*, and does not link the corresponding record.  This is confusing for the final user, since the line is visible, but disappears after saving...  So we change the implementation of this case to always link the updated record.

rco-odoo commented:
@odony @KangOl  can you review this for me? See the description above for the issue that is solved.

KangOl commented:
What about returning both?: `[(4, id), (1, id, values)]`

odony commented:
Some comments: 
- isn't the `[(1, id, values)]` initially returned by the `onchange` code a rather abusive optimization, with unclear semantics? Perhaps the `(4, id)` could be extended to `(4, id, values)` to provide the record values along with the ID, in the value returned by an on_change, while keeping the semantics clear? That's a big change for a stable series, though.
- upon save, regardless of the values returned by the `onchange`, shouldn't the web client compare the original values against the values being saved, and do the right thing, with a `(4,id)` in addition to the `(1,id,values)`? Maybe that's the most cost-effective fix?
- I think the SQL queries to test for m2m inclusion could be done as one query: `INSERT INTO ... WHERE NOT EXISTS ( SELECT 1 ... )`. That is, if we really need to do those queries, e.g. in the implementation of `4/LINK_TO`

rco-odoo commented:
@odony @KangOl I revised completely the fix, and fixed the web client instead.

Note that the `(1, id, values)` commands returned by the `onchange` is the format we agreed upon; so I did not change that format.

The bug was actually in the web client: after `read()`ing the values of the returned line, it is supposed to purge the "dirty" values it got from the `onchange`, and keep only the ones that differ from what it has read. That part of the code was missing.

I also fixed another issue along the way: when a line is updated, one has to put both `(4, id)` and `(1, id, values)` in the call to `write()`, as @KangOl suggested. Otherwise, the line is not added to the relation. This could be simpler if the command `(1, id, values)` was doing an implicit `(4, id)`...

odony commented:
That looks good, thanks for the updated patch!
One precision, though: is the LINK_TO always sent upon save, even if the x2m line was already present in the original `from_read` values? It seems that the web client is in the best position to know whether that ID was already linked, and can spare the server the extra LINK_TO check if it is unnecessary.
That does not preclude concurrent modifications that could have disconnected the record in the meantime, but that rare corner case belongs to the "conflict detection/resolution" category, and should not be the concern of the normal protocol (forcefully re-connecting the disconnected record is not necessarily the right conflict resolution).

rco-odoo commented:
@odony this extra commit adds the LINK_TO only when necessary...

odony commented:
Looks good, thanks!
